server/s3_list_object_versions_action.cc

```diff
@@ -0,0 +1,835 @@
+      if (prefix_match) {
+        delimiter_pos =
+            kv.first.find(request_delimiter, request_prefix.length());
+        if (delimiter_pos == std::string::npos) {
+          add_object_version(kv.first, kv.second.second);
+        } else {
+          // Roll up
+          // All keys under a common prefix are counted as a single key
+          // When we encounter a key that belongs to common prefix, the code
+          // maintains a state 'last_key_in_common_prefix', indicating a key
+          // with common prefix.
+          s3_log(S3_LOG_DEBUG, request_id,
+                 "Delimiter %s found at pos %zu in string %s\n",
+                 request_delimiter.c_str(), delimiter_pos, kv.first.c_str());
+          std::string common_prefix = kv.first.substr(0, delimiter_pos + 1);
+          skip_remaining_common_prefixes = false;
+          // Before adding common prefix, check if this is the first time we
+          // add this key to common_prefixes.
+          // If so, skip remaining keys that belong to it.
+          if (!is_prefix_in_common_prefix(common_prefix)) {
+            skip_remaining_common_prefixes = true;
+          }
+          if (common_prefix != request_key_marker) {
+            // If marker is specified, and if this key gets rolled up
+            // in common prefix, we add to common prefix only if it is not the
+            // same as specified marker
+            add_common_prefix(common_prefix);
+            s3_log(S3_LOG_DEBUG, request_id, "Adding common prefix [%s]\n",
+                   common_prefix.c_str());
+            last_common_prefix = common_prefix;
+            last_key_in_common_prefix = true;
+          }
+          if (skip_remaining_common_prefixes) {
+            // Skip remaining common prefixes
+            // For this, set last key for the next key enumeration and break
+            // Motr will skip returning the keys with the common prefix.
+            last_key = common_prefix + "\xff";
+            s3_log(S3_LOG_DEBUG, request_id,
+                   "Skipping further common prefixes, set next key = [%s]\n",
+                   last_key.c_str());
+            break;
+          }
+        }
+      } else {
+        // Prefix does not match.
+        // Check if fetched key is lexicographically greater than prefix
+        if (kv.first > request_prefix) {
+          // No further prefix match will occur (as items in Motr storage are
+          // arranaged in lexical order)
+          no_further_prefix_match = true;
+          // Set length to zero to set truncation to false
+          length = 0;
+          s3_log(S3_LOG_INFO, stripped_request_id,
+                 "No further prefix match. Skipping further object versions "
+                 "listing\n");
+          break;
+        }
+      }
+    }
+
+    if ((--length == 0) ||
+        (!last_key_in_common_prefix &&
+         ((versions_list.size() + common_prefixes.size()) == max_keys))) {
```

Comment:
Can this result in a value > max_keys?

Reply:
No, since we check this for every iteration of this "for" loop. And we break-out as soon as "(versions_list.size() + common_prefixes.size()) == max_keys)"